### PR TITLE
Convert crop from a clone-like operation into a transform

### LINKED ISSUE #115: Clone tool fails with Crop tool.
The clone tool doesn't work with cropped data. Cropped data needs to be a data transform in the pipeline.

### PR DESCRIPTION
**Pull request by mathturtle**
This fixes #115.  @Hovden how is this?

### CODE REVIEW

**Comment by Hovden** [comment]
After cropping, when I double click on the transform in the Data Transforms pipeline it crashes (OS X)

**Comment by Hovden** [comment]
Wouldn't it be easiest and most transparent to pass the crop bounds into a python transform?

It looks like a lot of infrastructure is being added to incorporate c routines in the data pipeline, this could be valuable for more complex routines, but cropping is a fast one time process. Users can then view and modify code.

**Comment by mathturtle** [comment]
@Hovden Yes it probably would be easiest to make it a Python transform.  But I'd like to make sure that it is possible to create C++ transforms and have them work properly too and this is the only good test case I have right now.

@cryos Would it make sense to add a new virtual function to Operator that throws up an 'Edit this Operator' dialog?  Right now the problem is that it is hard-coded to try to throw up the EditPythonOperatorDialog when a transform is double clicked without first checking that the operator is an OperatorPython.

**Comment by cryos** [comment]
I think in the case of the crop tool it is an ideal candidate for generalizing out operators to C++ *and* Python, something that will serve us well going forward. Any person using this would much rather see a dialog with the current crop, and the widget, all far more easily achieved from C++. I would be pretty unhappy to crop initially using a simple dialog, and a graphical widget, but when I go to edit being presented with Python code to edit.

Adding some flexibility here will make it easier when we want more tools that use graphical widgets in the 3D displays and/or custom dialogs for interaction. I don't think it is that much work to make one or two functions virtual, use them to get a dialog which can continue being a Python edit dialog in the case of Python operations, but could be any custom widget in the case of C++.

**Comment by mathturtle** [comment]
Ok, this branch got a little bigger than I expected.  I refactored the edit-this-operator code to use a widget generated by the operator that knows how to apply that operator's edit functionality.

**Comment by mathturtle** [comment]
This also adds an 'Apply' button to the operator edit dialog, so that changes can be applied and viewed without closing the dialog.

**Comment by mathturtle** [comment]
@cryos This is probably still a bit of a mess, but it works.

File: tomviz/CropOperator.cxx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+#include "vtkExtractVOI.h"
+#include "vtkImageData.h"
+#include "vtkNew.h"
+
+namespace tomviz
+{
+
+CropOperator::CropOperator(const int *dataExtent, const double *dataOrigin,
+                           const double *dataSpacing, QObject* prnt)
+  : Superclass(prnt)
```
Comment (cryos) [inline tomviz/CropOperator.cxx]: Really minor, but this seems ambiguous and I would rather we just used the supershort 'p' for parent. It is common in many projects, and this makes me think of print, and wonder what you are printing.

File: tomviz/CropOperator.cxx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+{
+  for (int i = 0; i < 6; ++i)
+    {
```
Comment (cryos) [inline tomviz/CropOperator.cxx]: Let's stick with modern indentation, we are using the standard C++ indentation scheme and it is inconsistently applied in this file.

File: tomviz/CropWidget.cxx [hunk 3]
```diff
@@ -45,24 +56,57 @@ class CropWidget::CWInternals
...
+  CropOperator *op;
+
+  Ui::CropWidget ui;
+  int dataExtent[6];
+  double dataOrigin[3];
+  double dataSpacing[3];
+  vtkBoundingBox dataBoundingBox;
+
+  void bounds(int bs[6]) const
+    {
```
Comment (cryos) [inline tomviz/CropWidget.cxx]: Use modern indentation, keep consistent with other functions.

File: tomviz/CropWidget.cxx [hunk 4]
```diff
@@ -45,24 +56,57 @@ class CropWidget::CWInternals
...
+    bs[index++] = this->ui.startX->value();
+    bs[index++] = this->ui.endX->value();
+    bs[index++] = this->ui.startY->value();
+    bs[index++] = this->ui.endY->value();
+    bs[index++] = this->ui.startZ->value();
+    bs[index++] = this->ui.endZ->value();
+    }
+
+  void blockSpinnerSignals(bool block)
+    {
```
Comment (cryos) [inline tomviz/CropWidget.cxx]: Here too

File: tomviz/CropWidget.cxx [hunk 5]
```diff
@@ -45,24 +56,57 @@ class CropWidget::CWInternals
...
-  vtkTrivialProducer *t = vtkTrivialProducer::SafeDownCast(
-    source->producer()->GetClientSideObject());
-  this->Internals->imageData = vtkImageData::SafeDownCast(t->GetOutputDataObject(0));
+  source->inputDataExtent(this->Internals->dataExtent);
+  source->inputDataOrigin(this->Internals->dataOrigin);
+  source->inputDataSpacing(this->Internals->dataSpacing);
 
+  double bounds[6];
+  for (int i = 0; i < 6; ++i)
+    {
```
Comment (cryos) [inline tomviz/CropWidget.cxx]: Here too

File: tomviz/CropWidget.cxx [hunk 6]
```diff
@@ -123,4 +203,58 @@ void CropWidget::updateBounds(int* boxBounds)
...
+void CropWidget::updateBounds(double *newBounds)
+{
+  Ui::CropWidget& ui = this->Internals->ui;
+
+  this->Internals->blockSpinnerSignals(true);
+
+  vtkBoundingBox newBoundingBox(newBounds);
+
+  if (this->Internals->dataBoundingBox.Intersects(newBoundingBox))
+    {
```
Comment (cryos) [inline tomviz/CropWidget.cxx]: Here too

File: tomviz/CropWidget.cxx [hunk 7]
```diff
@@ -123,4 +203,58 @@ void CropWidget::updateBounds(int* boxBounds)
...
+    ui.startY->setValue(vtkMath::Round(newBounds[2]));
+    ui.startZ->setValue(vtkMath::Round(newBounds[4]));
+
+    ui.endX->setValue(vtkMath::Round(newBounds[1]));
+    ui.endY->setValue(vtkMath::Round(newBounds[3]));
+    ui.endZ->setValue(vtkMath::Round(newBounds[5]));
+    }
+  // If there is no intersection use data extent
+  else
+    {
```
Comment (cryos) [inline tomviz/CropWidget.cxx]: Here too

File: tomviz/CropWidget.h [hunk 8]
```diff
@@ -16,41 +16,35 @@
...
-  void updateBounds(int* boxBounds);
+  virtual void applyChangesToOperator();
 
 private slots:
   void interactionEnd(vtkObject* caller);
-
-signals:
-  void bounds(double *bounds);
+  void valueChanged();
+  void updateBounds(int* boxBounds);
```
Comment (cryos) [inline tomviz/CropWidget.h]: * with variable name, consistent with updateBounds below

File: tomviz/EditOperatorDialog.cxx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+  Internals (new EditOperatorDialog::EODInternals())
+{
+  Q_ASSERT(o);
+  this->Internals->Op = o;
+  this->Internals->dataSource = dataSource;
+  this->Internals->needsToBeAdded = (dataSource != NULL);
+
+  QVariant position = this->Internals->loadPosition();
+  if (!position.isNull())
+    {
```
Comment (cryos) [inline tomviz/EditOperatorDialog.cxx]: Indentation

File: tomviz/EditOperatorDialog.cxx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+QSharedPointer<Operator>& EditOperatorDialog::op()
+{
+  return this->Internals->Op;
+}
+
+void EditOperatorDialog::onApply()
+{
+  this->Internals->Widget->applyChangesToOperator();
+  if (this->Internals->needsToBeAdded && this->Internals->dataSource != NULL)
+    {
```
Comment (cryos) [inline tomviz/EditOperatorDialog.cxx]: Indentation

File: tomviz/EditOperatorWidget.cxx [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  limitations under the License.
+
+******************************************************************************/
+
+#include "EditOperatorWidget.h"
+
+namespace tomviz
+{
+
+EditOperatorWidget::EditOperatorWidget(QWidget* prnt)
```
Comment (cryos) [inline tomviz/EditOperatorWidget.cxx]: Let's just use p

File: tomviz/EditOperatorWidget.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+// This class is the GUI needed to edit the properties of an operator.  The
+// operator will return one of these from its getEditorContents and it will
+// be shown in a dialog.  When Apply or Ok is clicked on the dialog, the
+// applyChangesToOperator() slot will be called on the widget.
+class EditOperatorWidget : public QWidget
+{
+  Q_OBJECT
+  typedef QWidget Superclass;
+public:
+  EditOperatorWidget(QWidget* prnt);
```
Comment (cryos) [inline tomviz/EditOperatorWidget.h]: Just use p

File: tomviz/OperatorPython.cxx [hunk 13]
```diff
@@ -87,6 +92,37 @@ namespace
...
     }
+
+  class EditPythonOperatorWidget : public tomviz::EditOperatorWidget
+  {
+    Q_OBJECT
+    typedef tomviz::EditOperatorWidget Superclass;
+  public:
+    EditPythonOperatorWidget(QWidget *prnt, tomviz::OperatorPython *o)
+      : Superclass(prnt), Op(o), Ui()
+      {
```
Comment (cryos) [inline tomviz/OperatorPython.cxx]: Intentation in this file

**Comment by cryos** [comment]
Minor comments inline. High level comment on editing the crop - it doesn't restore the current crop values and feels unexpected. I would like to see the widget and the dialog restore the last values. Tougher, and maybe one for the future, is restricting the crop widget to the bounds of the input data.

**Comment by cryos** [comment]
Overall I really like the changes, I would like to get this merged soon and then move on to some of the UI tweaks discussed.

**Comment by Hovden** [comment]
I like the crop widget to be unbounded. I find it easier to move around quickly and the behavior acts as expected.

Restoring the last used crop values is important UX.

Tested and runs well on OSX.

**Comment by mathturtle** [comment]
New version pushed.  I think I got all the indentation and variable name changes you highlighted.  The new version also restores the current crop values in the widget/dialog.

**Comment by cryos** [comment]
LGTM

**Pull request merged**